### PR TITLE
Implement TAP 4 (multiple repository consensus on entrusted targets) - WIP

### PR DESCRIPTION
**Pull request by vladimir-v-diaz**
**Fixes issue #**: #527

**Description of the changes being introduced by the pull request**:

This is an implementation of [TAP 4](https://github.com/theupdateframework/taps/blob/master/tap4.md) (multiple repository consensus on entrusted targets).  A MultiRepoUpdater class is added that reads `map.json`, determines which repository URL can be used for a requested target file, and returns the updater object(s) and targetinfo for the requested target file.  The client is free to use any of the returned updater objects to download the target file corresponding to the targetinfo.  An example is provided below:

```Python
multi_repo_updater = tuf.client.updater.MultiRepoUpdater("path/to/map.json")
valid_targetinfo = multi_repo_updater.get_valid_targetinfo(
    "foo.tgz", match_custom_field=False)

# `valid_targetinfo` contains one or more updaters that can be used to download
# the expected file for `foo.tgz.`  `match_custom_field` is a Boolean that indicates
# whether the optional custom field in targetinfo should match across the targetinfo
# provided by the threshold of repositories.

# All of the updaters in `valid_targetinfo` provide matching and trusted
# versions of `foo.tgz.`  Try each updater in `valid_targetinfo` until the client
# successfully downloads `foo.tgz.`
for my_updater, my_targetinfo = valid_targetinfo.iteritems():
  try:
    # Save `foo.tgz` to `path/to/destination.`
    my_updater.download_target(my_targetinfo, "path/to/destination")
    break
  
  except tuf.exceptions.NoWorkingMirrorError:
    continue
```

The client can also access the updater objects for each repository...
```Python
pypi_updater = multi_repo_updater.repository_names_to_updaters['pypi']
pypi_updater.download_target(...)
```

or

```Python
pypi_updater = multi_repo_updater.get_updater('pypi', map_file['repositories'])
pypi_updater.download_target(...)
```

@awwad PIease have a look at this PR and let me know if it supports everything needed to create Uptane repositories.  The repository tools now support the creation of multiple repositories within the same Python process.

**Please verify and check that the pull request fulfills the following
requirements**:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature

### CODE REVIEW

**Comment by vladimir-v-diaz** [comment]
@lukpueh @awwad 
I am done refactoring the TAP 4 implementation and would like for both you to review it.  You can safely ignore the large amount of JSON metadata that was added for testing.

File: tuf/client/updater.py [hunk 1]
```diff
@@ -145,6 +145,347 @@
...
+            'confined_target_dirs': ['']}
+
+        try:
+          # NOTE: State (e.g., keys) should NOT be shared across different
+          # updater instances.
+          logger.debug('Adding updater for ' + repr(repository_name))
+          updater = tuf.client.updater.Updater(repository_name, mirrors)
+          #updater.refresh()
+
+        except:
```
Comment (lukpueh) [inline tuf/client/updater.py]: Our [code style guidelines](https://github.com/secure-systems-lab/code-style-guidelines#use-class-based-exceptions) advice against bare `except:`s

**Comment by awwad** [comment]
Still working on the review, but I've added some issues about the text of TAP 4 here: [taps:90](https://github.com/theupdateframework/taps/issues/90), [taps:91](https://github.com/theupdateframework/taps/issues/91), [taps:92](https://github.com/theupdateframework/taps/issues/92), [taps:93](https://github.com/theupdateframework/taps/issues/93).

**Comment by vladimir-v-diaz** [comment]
I edited the summary to document the modified API for `get_valid_targetinfo().`

`multi_repo_updater.get_valid_targetinfo("foo.tgz", match_custom_field=True)` now returns a dictionary that contains the updaters that provide matching targetinfo for `foo.tgz.`

```
valid_targetinfo = {updater1: TARGETINFO, updater2: TARGETINFO, ...}
```

**Pull request merged**